csr_check.go

```diff
@@ -162,3 +180,77 @@ func authorizeCSR(machineList *v1beta1.MachineList, req *certificatesv1beta1.Cer
 
 	return nil
 }
+
+func authorizeNodeClientCSR(machines []v1beta1.Machine, nodes corev1client.NodeInterface, req *certificatesv1beta1.CertificateSigningRequest, csr *x509.CertificateRequest) error {
+	if !isReqFromNodeBootstrapper(req) {
+		return fmt.Errorf("CSR %s for node client cert has wrong user", req.Name)
+	}
+
+	nodeName := strings.TrimPrefix(csr.Subject.CommonName, nodeUserPrefix)
+	if len(nodeName) == 0 {
+		return fmt.Errorf("CSR %s has empty node name", req.Name)
+	}
+
+	_, err := nodes.Get(nodeName, metav1.GetOptions{})
+	switch {
+	case err == nil:
+		return fmt.Errorf("node %s already exists", nodeName)
+	case errors.IsNotFound(err):
+		// good, node does not exist
+	default:
+		return fmt.Errorf("failed to check if node %s already exists: %v", nodeName, err)
+	}
+
+	nodeMachine, ok := findMatchingMachineFromInternalDNS(nodeName, machines)
+	if !ok {
+		return fmt.Errorf("failed to find machine for node %s", nodeName)
+	}
+
+	if nodeMachine.Status.NodeRef != nil {
+		return fmt.Errorf("machine for node %s already has node ref", nodeName)
+	}
+
+	if !isWorkerMachine(nodeMachine) {
```

Comment:
I understand we don't want masters to go through this however is this effectively dropping machines which does not contain https://github.com/openshift/cluster-machine-approver/pull/26/files#diff-a482a1a664767249971b0b71ca9baaf0R28? How about any additional user machines/machineSets? I wonder if we might want to have an specific label for this e.g `machine.openshift.io/CSR: true` or possible better the other way around and have masters to opt out with `machine.openshift.io/CSR-need-approval: false`

Reply:
I wanted the most strict whitelist based approach.  This approval flow is only meant for workers so this check seemed appropriate.  I am not for or against adding a new label, we just need to discuss that type of change with the other stakeholders.